### PR TITLE
update performance and loss converging results

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #800

### CODE REVIEW

File: docs/performance.md [hunk 1]
```diff
@@ -1,39 +1,68 @@
-To demonstrate the effectiveness of PyTorch distributed training techniques used in torchtitan, we report both the infra metrics and loss curves of Llama 3 (8B and 70B) training on 64 A100 (80GB memory) GPUs and Llama 3.1 (405B) on 128 H100 (94GB memory).
-We report infra metrics achieved by [FSDP2](fsdp.md) (1D parallelism) under various configurations, and loss curves for both 1D parallelism (FSDP2) and 2D parallelism (FSDP2 + Tensor Parallel) training. (We only report 2D for 405B)
+We demonstrate the effectiveness of elastic distributed training using torchtitan, via experiments on Llama 3.1 8B, 70B, and 405B models, from 1D parallelism to 4D parallelism, at the scale from 8 GPUs to 512 GPUs.
 
+The experiments are conducted on NVIDIA H100 GPUs[^1] with 95 GiB memory, where each host is equipped with 8 GPUs and NVSwitch. Two hosts form a rack connected to a TOR switch. A backend RDMA network connects the TOR switches.
```
Comment (tianyu-l) [inline docs/performance.md]: In the recent Meta CP paper (https://arxiv.org/abs/2411.01783), they mentioned:
> We ran our performance benchmarks on the Grand Teton platform (Meta Engineering, 2022), where each host has 8 Nvidia H100 GPUs fully connected with NVLink (“host” and “node” are interchangeable in the subsequent text). Each H100 GPU is equipped with 96GB HBM2e with 2.4 TB/sec peak memory bandwidth. We tested on two subtypes of Grand Teton platforms: Grand Teton Training (GTT) and Grand Teton Inference (GTI). GTT hosts are inter-connected with backend RDMA network with 400 Gb/s per GPU, and GTI hosts are inter-connected with frontend network over TCP/IP with 100 Gb/s per GPU.

Would like to hear your thought on this @yifuwang

File: docs/performance.md [hunk 2]
```diff
@@ -1,39 +1,72 @@
...
-[^2]: Since for Float8, we are not converting all the matmuls to Float8 because our fused attention implementation is not done in Float8, so this number is lower than expected.
+| Parallelism | Sequence Length | TPS/GPU | Memory(GiB) |
+| ----- | ----: | ----: | ----: |
+| FSDP 8, CP 1 | 32768 | 76 | 75.3 |
+| FSDP 4, CP 2 | 65536 | 47 | 75.9 |
+| FSDP 2, CP 4 | 131072 | 31 | 77.1 |
+| FSDP 1, CP 8 | 262144 | 16 | 84.9 |
+
+
+#### Versions
```
Comment (lessw2020) [inline docs/performance.md]: minor...maybe "Versions Used For Performance Testing: "
It's overall obvious imo, but making it extra clear what Versions refers to exactly, might ensure no issues.

**Review by lessw2020 — APPROVED** [review]
Nice job - this is an improvement over the previous pages.

File: README.md [hunk 3]
```diff
@@ -52,7 +52,7 @@ Our guiding principles when building `torchtitan`:
     - estimate FSDP/HSDP memory usage without materializing the model
     - run distributed inference with Tensor Parallel
 
-We report our [Performance](docs/performance.md) verified on 64/128 GPUs.
+We report [performance](docs/performance.md) on up to 512 GPUs, and verify [loss converging](docs/converging.md) correctness of various techniques.
```
Comment (fduwjj) [inline README.md:55]: Just to clarify, the convergence curve now all come from 512 GPU run right?

**Comment by fduwjj** [comment]
Thanks for the great effort! I am wondering why do we remove the 405B convergence curve, are we saying down the road, we only show convergency curve for small 8B model if we add more parallelism down the road?

File: README.md (hunk 3 above)
Comment (tianyu-l) [reply README.md:55]: The performance tests are on 8 to 512 GPUs; the converging tests are on 8 to 64 GPUs. Please refer to `docs/converging.md` for details.

**Comment by tianyu-l** [comment]
> Thanks for the great effort! I am wondering why do we remove the 405B convergence curve, are we saying down the road, we only show convergency curve for small 8B model if we add more parallelism down the road? The reason I am asking this is because it is easier to see convergence issue in larger model than small ones. More broadly how did we decide 3000 steps is enough for 8B?

@fduwjj Thanks for the question. Let me try to explain.

For converging tests, there are two things of major interests: infra correctness and model/optimizer/lr_scheduler behaviors. It is sometimes hard to separate infra correctness and modeling correctness but we should try hard to.

As pointed out in https://github.com/pytorch/torchtitan/blob/main/docs/converging.md, in `converging.md` we only care about the infra side and believe FSDP to be correct (as the baseline). In this case, if some parallelism is not implemented correctly, most of the time the loss curve would show a (maybe tiny) gap pretty quick, without going for very long steps or to very large models. This has been the case for most loss debugging in the pass, but is not absolute, e.g. PP grad scaling bug (https://github.com/pytorch/pytorch/pull/144352) doesn't obviously affect the loss curves due to the robust behavior of AdamW (I feel this is true even for 405B).

So to directly answer your question, I think going to the largest possible model for infra-side testing is not necessary nor sufficient.

On the other hand, we need to make sure the models (and optimizer, lr scheduler combination) we add are "correct". But the difficulty is that it is hard to verify unless comparing with results outside as the correct "baseline". It is good if we run converging test for many steps and see if it's going well as a sanity check, but we can't identify any gap with baseline except for a loss explosion (which I think it sometimes feasible to bypass by simply tuning down learning rates, and again makes it "look right but still wrong"). So the value is there but kinda limited.

But still, I agree we probably should add such runs whenever we add new models (including your 405B 2D run) but under the model config folder.

**Comment by fduwjj** [comment]
@tianyu-l Got it. Thanks for your detailed explanation. I think this type of split (infra correctness vs. model/optimizer/lr_scheduler behaviors) makes it more clear when it comes to converges. So the bottemline here is that we need to ensure we don't have any infra incorrectness when bringing in new parallelism into titan and try our best to do due-diligence on the latter topic and ensure no obvious mistakes like loss curve explosion. Am I understanding this correctly?

> we probably should add such runs whenever we add new models (including your 405B 2D run) but under the model config folder.

Since you are here changing it, do you mind also including this change in this PR. I am ok with having a different PR doing it as well.

**Comment by tianyu-l** [comment]
@fduwjj 
> we need to ensure we don't have any infra incorrectness when bringing in new parallelism into titan and try our best to do due-diligence on the latter topic and ensure no obvious mistakes like loss curve explosion

I think this summarizes it well.

> do you mind also including this change in this PR. I am ok with having a different PR doing it as well.

Let's defer it in another PR -- I think it'd be enough if we only have the FSDP+TP run, with pytorch/torchtitan commits/versions/date documented as in this PR.
